weather_sp/splitter_pipeline/file_splitters.py

```diff
@@ -90,89 +82,123 @@ def should_skip(self):
                 if key not in outputs:
-                    metrics.Metrics.counter('file_splitters',
-                                            f'grib: {key}').inc()
-                    outputs[key] = self._open_outfile(key)
+                    outputs[key] = FileSystems.create(key)
                 outputs[key].write(grb.tostring())
                 outputs[key].flush()
 
             for out in outputs.values():
                 out.close()
-            self.logger.info('split %s into %d files', self.input_path, len(outputs))
+            self.logger.info('split %s into %d files',
+                             self.input_path, len(outputs))
 
     @contextmanager
     def _open_grib_locally(self) -> t.Iterator[t.Iterator[pygrib.gribmessage]]:
         with self._copy_to_local_file() as local_file:
             yield pygrib.open(local_file.name)
 
-    def _open_outfile(self, key: SplitKey):
-        return FileSystems.create(self._get_output_file_path(key))
-
 
 class NetCdfSplitter(FileSplitter):
 
+    def __init__(self, input_path: str, output_info: OutFileInfo,
+                 force_split: bool = False, logging_level: int = logging.INFO):
+        super().__init__(input_path, output_info,
+                         force_split, logging_level)
+
     def split_data(self) -> None:
+        if not self.output_info.split_dims():
+            raise ValueError('No splitting specified in template.')
         if self.should_skip():
             metrics.Metrics.counter('file_splitters', 'skipped').inc()
-            self.logger.info('Skipping %s, file already split.', repr(self.input_path))
+            self.logger.info('Skipping %s, file already split.',
+                             repr(self.input_path))
             return
-
-        with self._open_dataset_locally() as nc_data:
-            fields = [var for var in nc_data.variables.keys() if
-                      var not in nc_data.dimensions.keys()]
-            for field in fields:
-                self._create_netcdf_dataset_for_variable(nc_data, field)
-            self.logger.info('split %s into %d files', self.input_path, len(fields))
+        if any(split not in ('time', 'level', 'variable') for split in self.output_info.split_dims()):
+            raise ValueError(
+                'netcdf split: unknown split dimension, supported are time, level, variable')
+
+        with self._open_dataset_locally() as dataset:
+            if any(split not in dataset.dims and split not in ('variable') for split in self.output_info.split_dims()):
+                raise ValueError(
+                    'netcdf split: requested dimension not in dataset')
+            iterlists = []
+            iterlists.append([dataset[var].to_dataset(
+            ) for var in dataset.data_vars] if 'variable' in self.output_info.split_dims() else [dataset])
+            for dim in ('time', 'level'):
+                if dim in self.output_info.split_dims():
+                    iterlists.append(dataset[dim])
```

Comment:
These two may be coordinate indexes. Check out this method: 
https://github.com/google/weather-tools/blob/0ee97b4502f63fae35f605511c3e468b3ac6d2fa/weather_mv/loader_pipeline/util.py#L92

Specifically this line: 
https://github.com/google/weather-tools/blob/0ee97b4502f63fae35f605511c3e468b3ac6d2fa/weather_mv/loader_pipeline/util.py#L106

This may be useful to you to get the allowed coordinate dimensions to iterate over.

Reply:
I'll try to make NetCDF splitting more flexible in a follow-up PR if that's ok.